docker/Dockerfile

```diff
@@ -116,39 +120,136 @@ COPY --from=openssl_image /usr/local/lib /usr/local/lib
 RUN ldconfig \
  && ln -s /etc/ssl/certs/* /usr/local/ssl/certs/
 
-COPY VERSION /project/avalon/
-COPY ./bin /project/avalon/bin
+ENV TCF_HOME=/project/avalon
+
 COPY ./common/cpp /project/avalon/common/cpp
-COPY ./common/python /project/avalon/common/python
-COPY ./common/crypto_utils /project/avalon/common/crypto_utils
-COPY ./sdk /project/avalon/sdk
 
 WORKDIR /project/avalon/common/cpp
 
-RUN echo "Building Avalon Common CPP\n" \
-  && mkdir -p build \
+RUN mkdir -p build \
   && cd build \
   && cmake .. -DUNTRUSTED_ONLY=1 \
   && make
 
+
+
+# -------------=== common/python build ===-------------
+
+#Build common/cpp intermediate docker image
+FROM python_image as common_python_image
+
+RUN apt-get update \
+ && apt-get install -y -q \
+    make
+
+# Copy openssl build artifacts from openssl_image
+COPY --from=openssl_image /usr/local/ssl /usr/local/ssl
+COPY --from=openssl_image /usr/local/bin /usr/local/bin
+COPY --from=openssl_image /usr/local/include /usr/local/include
+COPY --from=openssl_image /usr/local/lib /usr/local/lib
+
+RUN ldconfig \
+ && ln -s /etc/ssl/certs/* /usr/local/ssl/certs/
+
+COPY VERSION /project/avalon/
+COPY ./bin /project/avalon/bin
+
+ENV TCF_HOME=/project/avalon
+
+COPY ./common/python /project/avalon/common/python
+
 WORKDIR /project/avalon/common/python
 
 RUN echo "Building Avalon Common Python\n" \
-  && make
+ && make
+
+
+
+# -------------=== common/crypto_utils build ===-------------
+
+#Build common/cpp intermediate docker image
+FROM python_image as common_crypto_utils_image
+
+RUN apt-get update \
+ && apt-get install -y -q \
+    swig \
+    make
+
+
+# Copy openssl build artifacts from openssl_image
+COPY --from=openssl_image /usr/local/ssl /usr/local/ssl
+COPY --from=openssl_image /usr/local/bin /usr/local/bin
+COPY --from=openssl_image /usr/local/include /usr/local/include
+COPY --from=openssl_image /usr/local/lib /usr/local/lib
+
+RUN ldconfig \
+ && ln -s /etc/ssl/certs/* /usr/local/ssl/certs/
+
+COPY --from=common_cpp_image /project/avalon/common/cpp/build/libuavalon-base64.a /project/avalon/common/cpp/build/
+COPY --from=common_cpp_image /project/avalon/common/cpp/build/libuavalon-common.a /project/avalon/common/cpp/build/
+COPY --from=common_cpp_image /project/avalon/common/cpp/build/libuavalon-crypto.a /project/avalon/common/cpp/build/
+COPY --from=common_cpp_image /project/avalon/common/cpp/build/libuavalon-parson.a /project/avalon/common/cpp/build/
+COPY --from=common_cpp_image /project/avalon/common/cpp/build/libuavalon-verify-ias-report.a /project/avalon/common/cpp/build/
```

Comment:
This looks bit ugly. Can we copy entire lib folder from common/cpp. I guess that would need creating a lib folder in commom/cpp build to copy build artifacts.